# 🌻 PRE-ENTREGA:
NOMBRE Y APELLIDO: **CECILIA FONTANA**


---





## 🔆 PRIMERA ETAPA: Recopilación y Preparación de Datos


##🔹1-Crear un documento en Google Colaboratory y cargar los sets de datos como DataFrames

### ✍*IMPORTAR* LAS LIBRERIAS *PANDAS - NUMPY*:

In [3]:
import pandas as pd
import numpy as np

#### *Montar la Unidad

In [4]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


### *Importar y listar datasets desde Google Drive.


In [7]:
import os
os.listdir("/content/drive/MyDrive/datasets")#Listamos los archivos que contiene la carpeta datasets elegida.

['marketing.csv', 'ventas.csv', 'clientes.csv']

### *Defino rutas relativas de los DataSets

In [8]:
ruta_ventas = "/content/drive/MyDrive/datasets/ventas.csv"
ruta_clientes = "/content/drive/MyDrive/datasets/clientes.csv"
ruta_marketing = "/content/drive/MyDrive/datasets/marketing.csv"

### *Cargo los csv como Data Frames, utilizando el método `read_csv` de Pandas (pd)

In [9]:
ventas = pd.read_csv(ruta_ventas)
clientes = pd.read_csv(ruta_clientes)
marketing = pd.read_csv(ruta_marketing)

### *Valido con un `print` y el método `.shape` para chequear que se cargo correctamente y luego muestro de cada DataFrame, con `display() y .head()` las primeras 4 filas.

In [10]:
print("ventas.shape ->", ventas.shape)
print("clientes.shape ->", clientes.shape)
print("marketing.shape ->", marketing.shape)

print("VENTAS:")
display(ventas.head(4))
print("CLIENTES:")
display(clientes.head(4))
print("MARKETING:")
display(marketing.head(4))

ventas.shape -> (3035, 6)
clientes.shape -> (567, 5)
marketing.shape -> (90, 6)
VENTAS:


id_venta           producto   precio  cantidad fecha_venta  \
0       792  Cuadro decorativo   $69.94       5.0  02/01/2024   
1       811    Lámpara de mesa  $105.10       5.0  02/01/2024   
2      1156           Secadora   $97.96       3.0  02/01/2024   
3      1372           Heladera  $114.35       8.0  02/01/2024   

           categoria  
0         Decoración  
1         Decoración  
2  Electrodomésticos  
3  Electrodomésticos

CLIENTES:


id_cliente               nombre  edad         ciudad  ingresos
0           1      Aloysia Screase    44  Mar del Plata  42294.68
1           2  Kristina Scaplehorn    25        Posadas  24735.04
2           3       Filip Castagne    50    Resistencia  35744.85
3           4          Liuka Luard    39   Bahía Blanca  27647.96

MARKETING:


id_campanha         producto  canal  costo fecha_inicio   fecha_fin
0           74  Adorno de pared     TV   4.81   20/03/2024  03/05/2024
1           12           Tablet   RRSS   3.40   26/03/2024  13/05/2024
2           32  Lámpara de mesa  Email   5.54   28/03/2024  20/04/2024
3           21       Smartphone   RRSS   6.37   29/03/2024  16/05/2024

## 🔹 2- Introducción a Pandas: realizar un análisis exploratorio inicial de los DataFrames.

### *Genero una Función para poder realizar un análisis exploratorio EDA, de cada DataFrame
Utilizo métodos  `.info()`, `.describe()`, `.shape`, `.columns`, `.dtypes.`, `.isna`, `.sum()`


In [11]:
def eda(df, nombre):
    print(f"=== {nombre} ===")
    print("shape:", df.shape)
    print("columnas:", list(df.columns))
    print("dtypes:")
    print(df.dtypes)
    print("\nNulos por columna:")
    print(df.isna().sum())
    print("\nPrimeras filas:")
    display(df.head(5))
    print("\nDescribe (numérico):")
    display(df.describe(include='number'))
    print("-"*100)


### *Ejecuto funcion para DataFrame ventas


In [12]:
eda(ventas, "VENTAS (inicial)")

=== VENTAS (inicial) ===
shape: (3035, 6)
columnas: ['id_venta', 'producto', 'precio', 'cantidad', 'fecha_venta', 'categoria']
dtypes:
id_venta         int64
producto        object
precio          object
cantidad       float64
fecha_venta     object
categoria       object
dtype: object

Nulos por columna:
id_venta       0
producto       0
precio         2
cantidad       2
fecha_venta    0
categoria      0
dtype: int64

Primeras filas:


id_venta           producto   precio  cantidad fecha_venta  \
0       792  Cuadro decorativo   $69.94       5.0  02/01/2024   
1       811    Lámpara de mesa  $105.10       5.0  02/01/2024   
2      1156           Secadora   $97.96       3.0  02/01/2024   
3      1372           Heladera  $114.35       8.0  02/01/2024   
4      1546           Secadora  $106.21       4.0  02/01/2024   

           categoria  
0         Decoración  
1         Decoración  
2  Electrodomésticos  
3  Electrodomésticos  
4  Electrodomésticos


Describe (numérico):


id_venta     cantidad
count  3035.000000  3033.000000
mean   1499.851400     6.496538
std     866.465379     3.457250
min       1.000000     1.000000
25%     748.500000     3.000000
50%    1502.000000     7.000000
75%    2249.500000     9.000000
max    3000.000000    12.000000

----------------------------------------------------------------------------------------------------


## ✔ Analisis de los datos Iniciales del DataFrame ventas:


*   Tenemos **6** columnas y **3035** registros o filas
*   Hay datos que se deberán normalizar, como por ejemplo **fecha_venta** , **precio**, etc.


*   Tenemos **2 nulos en la columna precio** y **2 en la columna cantidad** que deberemos analizar y eliminar si es preciso.
*   Al describir los datos numéricos contamos con algunas estadísticas que luego nos ayudaran a trabajar con los datos cuando esten normalizados y limpios,





### *Ejecuto funcion para DataFrame clientes


In [13]:
eda(clientes, "CLIENTES (inicial)")

=== CLIENTES (inicial) ===
shape: (567, 5)
columnas: ['id_cliente', 'nombre', 'edad', 'ciudad', 'ingresos']
dtypes:
id_cliente      int64
nombre         object
edad            int64
ciudad         object
ingresos      float64
dtype: object

Nulos por columna:
id_cliente    0
nombre        0
edad          0
ciudad        0
ingresos      0
dtype: int64

Primeras filas:


id_cliente               nombre  edad         ciudad  ingresos
0           1      Aloysia Screase    44  Mar del Plata  42294.68
1           2  Kristina Scaplehorn    25        Posadas  24735.04
2           3       Filip Castagne    50    Resistencia  35744.85
3           4          Liuka Luard    39   Bahía Blanca  27647.96
4           5        Dore Cockshtt    28        Rosario  28245.65


Describe (numérico):


id_cliente        edad      ingresos
count  567.000000  567.000000    567.000000
mean   284.000000   37.940035  34668.739012
std    163.823075   10.202885  12974.531446
min      1.000000   20.000000    170.290000
25%    142.500000   30.000000  26015.240000
50%    284.000000   37.000000  35066.830000
75%    425.500000   43.000000  42457.100000
max    567.000000   81.000000  88053.010000

----------------------------------------------------------------------------------------------------


## ✔ Analisis de los datos Iniciales del DataFrame clientes:


*   Tenemos **5** columnas y **567** registros o filas
*   No se encuentran datos que se deberán normalizar en una primera instancia.

*   No Tenemos valores nulos que se deberían eliminar.
*   Al describir los datos numéricos contamos con algunas estadísticas que luego nos ayudaran a trabajar con los datos cuando esten normalizados y limpios.

### *Ejecuto funcion para DataFrame marketing


In [14]:
eda(marketing, "MARKETING (inicial)")

=== MARKETING (inicial) ===
shape: (90, 6)
columnas: ['id_campanha', 'producto', 'canal', 'costo', 'fecha_inicio', 'fecha_fin']
dtypes:
id_campanha       int64
producto         object
canal            object
costo           float64
fecha_inicio     object
fecha_fin        object
dtype: object

Nulos por columna:
id_campanha     0
producto        0
canal           0
costo           0
fecha_inicio    0
fecha_fin       0
dtype: int64

Primeras filas:


id_campanha         producto  canal  costo fecha_inicio   fecha_fin
0           74  Adorno de pared     TV   4.81   20/03/2024  03/05/2024
1           12           Tablet   RRSS   3.40   26/03/2024  13/05/2024
2           32  Lámpara de mesa  Email   5.54   28/03/2024  20/04/2024
3           21       Smartphone   RRSS   6.37   29/03/2024  16/05/2024
4           58         Alfombra  Email   4.25   31/03/2024  05/05/2024


Describe (numérico):


id_campanha      costo
count    90.000000  90.000000
mean     45.500000   4.928667
std      26.124701   0.947750
min       1.000000   2.950000
25%      23.250000   4.372500
50%      45.500000   4.900000
75%      67.750000   5.562500
max      90.000000   7.390000

----------------------------------------------------------------------------------------------------


## ✔ Analisis de los datos Iniciales del DataFrame marketing:


*   Tenemos **6** columnas y **90** registros o filas
*   Se encuentran datos que se deberán normalizar como los de la columna **fecha_inicio**, **fecha_fin**,  en una primera instancia.

*   No Tenemos valores nulos que se deberían eliminar.
*   Al describir los datos numéricos contamos con algunas estadísticas que luego nos ayudaran a trabajar con los datos cuando esten normalizados y limpios.

## 🔹 3- Calidad de Datos (nulos o duplicados)
Identificar valores nulos y duplicados en los conjuntos de datos. Documentar el estado inicial de los datos.

### *Genero una Función para poder realizar un control de la calidad de datos de cada DataFrame y que muestre:


1.   La cantidad de valores nulos por columna
2.   El total de filas completamente duplicadas.


1.  Si se especifica una columna clave, los valores mas repetidos de esa clave.

Utilizo métodos  `.isna`, `.sum()`, `.to_frame`, `.duplicates`, `.value_counts()`, `.sort_values`

In [15]:
def calidad(df, nombre, clave=None):

    print(f"### {nombre}")

    # -------------------------------------------------
    # Mostrar cantidad de valores nulos por columna
    # -------------------------------------------------
    display(df.isna().sum().to_frame("nulos"))

    # -------------------------------------------------
    # Contar filas duplicadas completas
    # -------------------------------------------------

    dup_rows = df.duplicated(keep=False).sum()
    print("Filas duplicadas (exactas):", dup_rows)

    # -------------------------------------------------
    # Si se especificó una columna clave válida, analizar duplicados por esa columna
    # -------------------------------------------------

    if clave and clave in df.columns:
      # Contar cuántas filas tienen valores repetidos en esa columna
        dup_key = df[clave].duplicated(keep=False).sum()
        print(f"Duplicados por clave '{clave}':", dup_key)

        # Si existen duplicados, mostrar cuáles son los valores más repetidos
        if dup_key > 0:

            duplicados_ordenados = (
                df[df[clave].duplicated(keep=False)][clave]
                .value_counts()
                .sort_values(ascending=False)
            )

            print("\n🔁 Top valores duplicados más frecuentes:")
            # Mostrar solo los primeros 10 (los más repetidos)
            display(duplicados_ordenados.head(10))
        else:
            print(f"No se encontraron duplicados en la clave '{clave}'.")
    else:
        # Si la clave no fue pasada o no existe en el DataFrame
        if clave:
            print(f"La clave '{clave}' no existe en el DataFrame.")
        else:
            print("No se indicó una clave para analizar duplicados por columna.")


### *Ejecuto funcion `calidad`para el DataFrame ventas por su clave `id_venta`


In [16]:
calidad(ventas, "VENTAS", clave="id_venta")


### VENTAS


nulos
id_venta         0
producto         0
precio           2
cantidad         2
fecha_venta      0
categoria        0

Filas duplicadas (exactas): 70
Duplicados por clave 'id_venta': 70

🔁 Top valores duplicados más frecuentes:


id_venta
56      2
421     2
424     2
1868    2
2545    2
2778    2
145     2
300     2
439     2
906     2
Name: count, dtype: int64

## ✔ Analisis de la calidad de los datos DataFrame ventas según su clave.

### *Ejecuto funcion `calidad`para el DataFrame ventas general, sin clave


In [17]:
calidad(ventas, "VENTAS")

### VENTAS


nulos
id_venta         0
producto         0
precio           2
cantidad         2
fecha_venta      0
categoria        0

Filas duplicadas (exactas): 70
No se indicó una clave para analizar duplicados por columna.


## ✔ Analisis de la calidad de los datos DataFrame ventas en general, sin indicar.

### *Ejecuto funcion `calidad`para el DataFrame clientes por su clave `id_cliente`


In [18]:
calidad(clientes, "CLIENTES", clave="id_cliente")


### CLIENTES


nulos
id_cliente      0
nombre          0
edad            0
ciudad          0
ingresos        0

Filas duplicadas (exactas): 0
Duplicados por clave 'id_cliente': 0
No se encontraron duplicados en la clave 'id_cliente'.


## ✔ Analisis de la calidad de los datos DataFrame clientes según su clave.


*   Encontramos que no hay filas duplicadas, según su clave.



### *Ejecuto funcion `calidad`para el DataFrame clientes en general, sin clave.


In [19]:
calidad(clientes, "CLIENTES")


### CLIENTES


nulos
id_cliente      0
nombre          0
edad            0
ciudad          0
ingresos        0

Filas duplicadas (exactas): 0
No se indicó una clave para analizar duplicados por columna.


## ✔ Analisis de la calidad de los datos DataFrame clientes en general, sin indicar.


*   No se encuentran ni nulos, ni filas duplicadas.



### *Ejecuto funcion `calidad`para el DataFrame marketing por su clave `id_campanha`


In [20]:
calidad(marketing, "MARKETING", clave="id_campanha")


### MARKETING


nulos
id_campanha       0
producto          0
canal             0
costo             0
fecha_inicio      0
fecha_fin         0

Filas duplicadas (exactas): 0
Duplicados por clave 'id_campanha': 0
No se encontraron duplicados en la clave 'id_campanha'.


## ✔ Analisis de la calidad de los datos DataFrame marketing en general, indicando su clave.


*   No se encuentran ni nulos, ni filas duplicadas.
Se decide no ejecutar funcion sin la clave, porque el resultado srá el mismo.



## 🔆 SEGUNDA ETAPA: Preprocesamiento y Limpieza de Datos


---





## 🔹 4- Limpieza de Datos: Limpiar el conjunto de datos  eliminando duplicados y caracteres no deseados.
 Documentar el proceso y los resultados.


## **Limpieza del dataset Y normalización del dataset**
- Eliminamos duplicados.
- Normalizamos **texto** en columnas `object` (trim + capitalización simple).
- Convertimos fechas a fechas reales
- Convertimos `precio` y `cantidad` a numéricos si existen.
- Guardamos CSV limpios.

## *Realizamos la  
##### ============================================
#####  🧹 LIMPIEZA Y NORMALIZACIÓN DE LOS DATASETS
##### ============================================
##### Se limpian y normalizan los DataFrames:
#####     ventas, clientes, marketing
##### ============================================

In [21]:
# -------------------------------------------------
# 1️⃣ Crear copias independientes para no modificar los originales
# -------------------------------------------------
ventas_clean = ventas.copy()
clientes_clean = clientes.copy()
marketing_clean = marketing.copy()


In [22]:

# -------------------------------------------------
# 2️⃣ Eliminar filas completamente duplicadas
# -------------------------------------------------
ventas_clean = ventas_clean.drop_duplicates()
clientes_clean = clientes_clean.drop_duplicates()
marketing_clean = marketing_clean.drop_duplicates()

### *Ejecuto funcion `calidad`para los DataFrames copia, luego de haber eliminado las filas completamente duplicadas, para verificar que no queden elementos con èstas caracterìsticas.


In [23]:
calidad(ventas_clean, "VENTAS CLEAN", clave="id_venta")

### VENTAS CLEAN


nulos
id_venta         0
producto         0
precio           2
cantidad         2
fecha_venta      0
categoria        0

Filas duplicadas (exactas): 0
Duplicados por clave 'id_venta': 0
No se encontraron duplicados en la clave 'id_venta'.


Sòlo lo realizamos con el dataset copia clientes, ya que ya hemos detectado con anterioridad que no tenemos duplicados en el los otros dos datasets

### *Genero funcion `normalizar_texto`para limpiar texto en columnas tipo String.


In [24]:
# -------------------------------------------------
# 3️⃣ Función para limpiar texto en columnas tipo string
# -------------------------------------------------
def normalizar_texto(df):
    for col in df.select_dtypes(include="object").columns:

        df[col] = (
            df[col]
            .astype(str)
            .str.strip()
            .str.replace(r"[\u200b\t\r\n]", "", regex=True)
            .str.replace(" +", " ", regex=True)
            .str.title()
        )
    return df

### *Genero algoritmo para normalizar las fechas.


---
Utilizo primero código para normalizar las fechas y luego recién ejecuto
funcion de texto.

In [25]:
# -------------------------------------------------
# Normalizar fechas
# -------------------------------------------------
for df in [ventas_clean, clientes_clean, marketing_clean]:
    for col in df.columns:
        if "fecha" in col.lower():  # detecta columnas con la palabra "fecha"
            df[col] = pd.to_datetime(df[col], errors="coerce", dayfirst=True)


* Corroboro que las fechas quedaron normalizadas

In [26]:
print(ventas_clean.dtypes)
print(clientes_clean.dtypes)
print(marketing_clean.dtypes)

id_venta                int64
producto               object
precio                 object
cantidad              float64
fecha_venta    datetime64[ns]
categoria              object
dtype: object
id_cliente      int64
nombre         object
edad            int64
ciudad         object
ingresos      float64
dtype: object
id_campanha              int64
producto                object
canal                   object
costo                  float64
fecha_inicio    datetime64[ns]
fecha_fin       datetime64[ns]
dtype: object


### *Ahora corremos la funcion para normalizar textos, e inmediatamente verificamos que asi sea.


In [27]:
#  Aplicar la normalización de texto
# -------------------------------------------------
ventas_clean = normalizar_texto(ventas_clean)
clientes_clean = normalizar_texto(clientes_clean)
marketing_clean = normalizar_texto(marketing_clean)

In [28]:
#mostramos los df luego de normalizar los textos para revisar que se haya ejecutado normalmente
print(ventas_clean.head(10))
print(clientes_clean.head(10))
print(marketing_clean.head(10))

   id_venta           producto   precio  cantidad fecha_venta  \
0       792  Cuadro Decorativo   $69.94       5.0  2024-01-02   
1       811    Lámpara De Mesa  $105.10       5.0  2024-01-02   
2      1156           Secadora   $97.96       3.0  2024-01-02   
3      1372           Heladera  $114.35       8.0  2024-01-02   
4      1546           Secadora  $106.21       4.0  2024-01-02   
5      1697    Horno Eléctrico   $35.35       9.0  2024-01-02   
6      1710   Plancha De Vapor   $65.43       2.0  2024-01-02   
7      2959          Proyector   $88.17       9.0  2024-01-02   
8       318  Rincón De Plantas   $79.86      11.0  2024-01-03   
9       419         Candelabro   $66.11       8.0  2024-01-03   

           categoria  
0         Decoración  
1         Decoración  
2  Electrodomésticos  
3  Electrodomésticos  
4  Electrodomésticos  
5  Electrodomésticos  
6  Electrodomésticos  
7        Electrónica  
8         Decoración  
9         Decoración  
   id_cliente               nom

### *Genero  algoritmo para Normalizar valores numéricos.

* Vamos con el campo, en este caso lo realizo con el campo `precio` .


In [29]:
# -------------------------------------------------
# 6️⃣ Normalizar valores numéricos
# -------------------------------------------------
# 🏷️ Campo "precio"
if "precio" in ventas_clean.columns:
    ventas_clean["precio"] = (
        ventas_clean["precio"]
        .astype(str)
        .str.replace("$", "", regex=False)
        .str.replace(",", "", regex=False)
        .str.strip()
    )
    ventas_clean["precio"] = pd.to_numeric(ventas_clean["precio"], errors="coerce")



* Imprimimos para verificar, los tipos de campo, las columnas y especifico por el campo precio, vemos los valores acordes.

In [30]:
print(ventas_clean.dtypes)
print(ventas_clean.columns)
print(ventas_clean["precio"].head(10))

id_venta                int64
producto               object
precio                float64
cantidad              float64
fecha_venta    datetime64[ns]
categoria              object
dtype: object
Index(['id_venta', 'producto', 'precio', 'cantidad', 'fecha_venta',
       'categoria'],
      dtype='object')
0     69.94
1    105.10
2     97.96
3    114.35
4    106.21
5     35.35
6     65.43
7     88.17
8     79.86
9     66.11
Name: precio, dtype: float64


* Vamos con el campo, en este caso lo realizo con el campo `cantidad` .


In [31]:
# 🧮 Campo "cantidad"
if "cantidad" in ventas_clean.columns:
    ventas_clean["cantidad"] = pd.to_numeric(
        ventas_clean["cantidad"], errors="coerce"
    ).astype("Int64")

* Imprimimos para verificar, que precio haya quedado tipo int64

In [32]:
print(ventas_clean.dtypes)

id_venta                int64
producto               object
precio                float64
cantidad                Int64
fecha_venta    datetime64[ns]
categoria              object
dtype: object


## ✔ Generamos nuevos archivos .csv para resguardar los datos filtrados, normalizados.
*   No es una operatoria necesaria, pero no es mala práctica guardarlos en un principio en archivos nuevos.



In [33]:
# -------------------------------------------------
# 7️⃣ Guardar los DataFrames limpios como CSV
# -------------------------------------------------
print(ventas_clean.head(10))
print(clientes_clean.head(10))
print(marketing_clean.head(10))
ventas_clean.info()
ventas_clean.to_csv("/content/drive/MyDrive/datasets/ventas_clean.csv", index=False)
clientes_clean.to_csv("/content/drive/MyDrive/datasets/clientes_clean.csv", index=False)
marketing_clean.to_csv("/content/drive/MyDrive/datasets/marketing_clean.csv", index=False)

print("✅ Archivos guardados: ventas_clean.csv, clientes_clean.csv, marketing_clean.csv")

   id_venta           producto  precio  cantidad fecha_venta  \
0       792  Cuadro Decorativo   69.94         5  2024-01-02   
1       811    Lámpara De Mesa  105.10         5  2024-01-02   
2      1156           Secadora   97.96         3  2024-01-02   
3      1372           Heladera  114.35         8  2024-01-02   
4      1546           Secadora  106.21         4  2024-01-02   
5      1697    Horno Eléctrico   35.35         9  2024-01-02   
6      1710   Plancha De Vapor   65.43         2  2024-01-02   
7      2959          Proyector   88.17         9  2024-01-02   
8       318  Rincón De Plantas   79.86        11  2024-01-03   
9       419         Candelabro   66.11         8  2024-01-03   

           categoria  
0         Decoración  
1         Decoración  
2  Electrodomésticos  
3  Electrodomésticos  
4  Electrodomésticos  
5  Electrodomésticos  
6  Electrodomésticos  
7        Electrónica  
8         Decoración  
9         Decoración  
   id_cliente               nombre  edad  

* Primero imprimimos los df para verificar y luego realizamos la copia que se determina realizada con el resultado de archivos guardados. Esto no modifica en absoluto los DataFrames guardados en memoria.
Agrega 3 archivos nuevos en la carpeta Datasets.

## ✔ Reporte Global luego de la limpieza de datos


*   Se genera una función `reporte_calidad_global` que lle los 3 data frames limpios y muestra un resumen comparativo de nulos, duplicados y tipos de datos.


In [34]:
# ============================================
# 📊 REPORTE GLOBAL DE CALIDAD DE DATOS
# ============================================

def reporte_calidad_global(dfs, nombres):
    resumen = []
    for df, nombre in zip(dfs, nombres):
        nulos = df.isna().sum().sum()
        duplicados = df.duplicated(keep=False).sum()
        columnas = len(df.columns)
        filas = len(df)

        resumen.append({
            "Dataset": nombre,
            "Filas": filas,
            "Columnas": columnas,
            "Nulos totales": nulos,
            "Duplicados": duplicados,
        })

    reporte = pd.DataFrame(resumen)
    #display(reporte)
    return reporte


##### ============================================
##### ✅ Ejemplo de uso
##### ============================================
* Se ejecuta la función con los df originales y con los limpios. Luego se verifican los resultados.

In [35]:
#Ejecutamos la funcion con los df originales
print(reporte_calidad_global([ventas, clientes, marketing], ["VENTAS Original", "CLIENTES Original", "MARKETING Original"]))
#Ejecutamos con los limpios
print(reporte_calidad_global([ventas_clean, clientes_clean, marketing_clean],["VENTAS Copia   ", "CLIENTES Copia   ", "MARKETING Copia   "]))


              Dataset  Filas  Columnas  Nulos totales  Duplicados
0     VENTAS Original   3035         6              4          70
1   CLIENTES Original    567         5              0           0
2  MARKETING Original     90         6              0           0
              Dataset  Filas  Columnas  Nulos totales  Duplicados
0     VENTAS Copia      3000         6              4           0
1   CLIENTES Copia       567         5              0           0
2  MARKETING Copia        90         6              0           0


*Acá verificamos que la limpieza y normalización fue realizada.

## 🔹 5- Transformación de Datos (filtrar "alto rendimiento").


## ✔ Objetivo buscado


*   Detectar los productos con mejor desempeño económico (top 20% por ingreso total)
* Aplicar transformación: calcular ingreso, agregar por producto y filtrar.

* Se genera una función `encontrar_columna`, que busca la columna cuyo nombre contenga alguno de los patrones dados.

In [36]:
def encontrar_columna(df, candidatos):
    # Recorremos todas las columnas del DataFrame
    for c in df.columns:

        # Convertimos el nombre de la columna a minúsculas
        # Esto se hace para comparar sin importar si está escrito con mayúsculas o minúsculas
        nombre = c.lower()
        if any(p in nombre for p in candidatos):
            # Si encontramos una coincidencia, devolvemos el nombre original de la columna
            return c
    # Si termina el bucle y no se encontró ninguna coincidencia, devolvemos None
    return None

*Hacemos correr la función para detectar la columna producto.

In [37]:
# 1️⃣👀 Detectar la columna de producto
prod_col = encontrar_columna(ventas_clean, ["producto", "id_producto", "sku", "articulo", "artículo"])
if prod_col is None:
    raise ValueError("No se encontró columna de producto. Renombrá una columna a 'producto' o similar.")

print(prod_col)


producto


* Calculamos el ingreso por registro, es decir el `precio*cantidad` y lo asignamos en una nueva columna llamada `ingreso`

In [40]:
# 2️⃣👀 Calcular ingreso por registro = precio * cantidad y asignamos en la columna ingreso
ventas_perf = (
    ventas_clean
    .assign(
        ingreso = ventas_clean["precio"] * ventas_clean["cantidad"]
    )
)

*Imprimimos el df ventas para verificar los datos

In [42]:
print (ventas_perf.columns)
display(ventas_perf.head(5))

Index(['id_venta', 'producto', 'precio', 'cantidad', 'fecha_venta',
       'categoria', 'ingreso'],
      dtype='object')


id_venta           producto  precio  cantidad fecha_venta  \
0       792  Cuadro Decorativo   69.94         5  2024-01-02   
1       811    Lámpara De Mesa  105.10         5  2024-01-02   
2      1156           Secadora   97.96         3  2024-01-02   
3      1372           Heladera  114.35         8  2024-01-02   
4      1546           Secadora  106.21         4  2024-01-02   

           categoria  ingreso  
0         Decoración    349.7  
1         Decoración    525.5  
2  Electrodomésticos   293.88  
3  Electrodomésticos    914.8  
4  Electrodomésticos   424.84

*Acá observamos en el nuevo data Frame `ventas_perf` que tenemos los datos anteriores mas la columna que se agrega y cuyo contenido es la operación pedida.

* Ahora trabajaremos con las métricas por producto

In [43]:
# 3️⃣ 👀Agregar métricas por producto
resumen_prod = (
    ventas_perf
    # 1) Agrupamos el DataFrame por una o varias columnas clave
    .groupby(
        by=prod_col,
        dropna=False,
        as_index=False,
        observed=False
    )
    # 2) Agregamos (resumimos) columnas numéricas por cada grupo
    .agg(
        ingreso_total=('ingreso', 'sum'),
        unidades=('cantidad', 'sum'),
        precio_promedio=('precio', 'mean'),
        registros=('ingreso', 'size')
    )
)

In [44]:
print(resumen_prod.head(60))
#ordenar resumen_prod por el mayor ingreso_total, y redondear precio_promedio a 2 decimales redondeado

                  producto  ingreso_total  unidades  precio_promedio  \
0          Adorno De Pared       48093.49       633        76.097800   
1                 Alfombra       44773.06       615        74.098300   
2               Aspiradora       50085.86       651        77.447400   
3              Auriculares       74175.58       958        76.302727   
4                 Batidora        50979.2       672        77.537200   
5                 Cafetera       59607.31       765        79.046581   
6               Candelabro        11128.8       160        74.595417   
7   Consola De Videojuegos       46174.41       623        76.346869   
8                 Cortinas       44865.03       610        73.649900   
9        Cuadro Decorativo        54297.6       726        74.578000   
10          Cámara Digital       45217.96       638        72.010700   
11   Elementos De Cerámica       45411.09       636        71.611515   
12       Espejo Decorativo       46783.31       635        75.66

In [48]:
resumen_prod = resumen_prod.sort_values(by='ingreso_total', ascending=False)
resumen_prod['precio_promedio'] = resumen_prod['precio_promedio'].round(2)
resumen_prod['ingreso_total'] = resumen_prod['ingreso_total'].round(2) #no se por que no me queda con dos decimales todo.
display(resumen_prod.head(60))

producto  ingreso_total  unidades  precio_promedio  \
19         Lámpara De Mesa       82276.38      1112            72.72   
3              Auriculares       74175.58       958            76.30   
20              Microondas       72562.89       912            79.18   
5                 Cafetera       59607.31       765            79.05   
9        Cuadro Decorativo        54297.6       726            74.58   
26              Smartphone       54132.44       665            81.40   
25                Secadora       52115.45       696            73.34   
16       Jarrón Decorativo       51130.88       672            74.53   
4                 Batidora        50979.2       672            77.54   
24       Rincón De Plantas       50456.45       691            74.82   
13      Freidora Eléctrica       50155.15       630            78.25   
2               Aspiradora       50085.86       651            77.45   
15         Horno Eléctrico        49913.9       633            75.69   
18                Lavadora       48946.44       671            74.95   
23               Proyector        48187.0       626            74.68   
0          Adorno De Pared       48093.49       633            76.10   
27              Smartwatch       47836.86       620            75.54   
21     Parlantes Bluetooth       47779.46       660            73.58   
29               Televisor       47038.04       632            73.82   
12       Espejo Decorativo       46783.31       635            75.66   
17                  Laptop       46179.48       617            75.33   
7   Consola De Videojuegos       46174.41       623            76.35   
11   Elementos De Cerámica       45411.09       636            71.61   
10          Cámara Digital       45217.96       638            72.01   
22        Plancha De Vapor       45196.63       632            73.04   
8                 Cortinas       44865.03       610            73.65   
1                 Alfombra       44773.06       615            74.10   
28                  Tablet       25856.57       374            71.38   
14                Heladera        25736.8       330            74.13   
6               Candelabro        11128.8       160            74.60   

    registros  
19        176  
3         143  
20        135  
5         117  
9         100  
26        101  
25        100  
16        100  
4         100  
24        101  
13        100  
2         100  
15        100  
18        100  
23         99  
0         100  
27        100  
21         99  
29        100  
12        100  
17        100  
7          99  
11        100  
10        100  
22        100  
8         100  
1         100  
28         58  
14         48  
6          24

* Ahora trabajaremos con el percentil 80. En este caso queremos saber el ingreso que separa al 80% de los productos con menores ingresos del 20% con mayores ingresos.

In [49]:
# 4️⃣ 👀Calcular percentil 80 de ingreso_total
# --------------------------------------------------------
p80_ingreso = resumen_prod["ingreso_total"].quantile(
    q=0.80,
    interpolation="linear"
)

# En resumen:
# - quantile calcula el valor límite de un percentil.
# - q define qué percentil queremos.
# - interpolation define cómo se calcula cuando el valor no está exactamente en los datos.
# El resultado (p80_ingreso) es el ingreso total que marca el límite superior del 80% de los productos.


* En resumen:
 - quantile calcula el valor límite de un percentil.
 - q define qué percentil queremos.
 - interpolation define cómo se calcula cuando el valor no está exactamente en los datos.
* El resultado (p80_ingreso) es el ingreso total que marca el límite superior del 80% de los productos.

In [50]:
# 5️⃣ 👀Filtrar los productos "de alto rendimiento" y ordenarlos
# -------------------------------------------------------------------
# Objetivo: quedarnos con los productos cuyo ingreso_total está en el 20% superior
# (ingreso_total >= p80_ingreso) y luego ordenarlos de mayor a menor por ingreso y unidades.
ventas_top = (
    resumen_prod
    .query(
        "ingreso_total >= @p80_ingreso",
        engine="python"

    )
    .sort_values(
        by=["ingreso_total", "unidades"],
        ascending=[False, False],
        na_position="last",
        ignore_index=False
    )
)


*Resultado:
* `ventas_top` contiene solo los productos cuyo ingreso_total >= p80_ingreso,
* ordenados de mayor a menor por ingreso_total y, ante empates, por unidades.

*Se muestran los resultados obtenidos

In [51]:
# 6️⃣ 👀 Mostrar resultados
print(f"Columna de producto detectada: {prod_col}")
print(f"Umbral (percentil 80) de ingreso_total: {float(p80_ingreso):,.2f}")
print("✅ Productos de ALTO RENDIMIENTO (top 20% por ingreso):")
display(ventas_top.head(60))

Columna de producto detectada: producto
Umbral (percentil 80) de ingreso_total: 52,518.85
✅ Productos de ALTO RENDIMIENTO (top 20% por ingreso):


producto  ingreso_total  unidades  precio_promedio  registros
19    Lámpara De Mesa       82276.38      1112            72.72        176
3         Auriculares       74175.58       958            76.30        143
20         Microondas       72562.89       912            79.18        135
5            Cafetera       59607.31       765            79.05        117
9   Cuadro Decorativo        54297.6       726            74.58        100
26         Smartphone       54132.44       665            81.40        101

## 🔹 6- Agregación(resumen por categoría y análisis de ingresos.).

## ✔ Objetivo buscado


*   Resumir las ventas por categoría de producto con métricas útiles (ingreso total, unidades, cantidad de ventas, ticket promedio) y analizar los ingresos generados.


* **Plan a seguir**:

---



** Detectar la columna de categoría.

** Asegurar columna ingreso (si no existe, crearla).

** groupby(categoría).agg(...) para obtener métricas.

** Ordenar por ingreso_total.

** Calcular ticket_promedio_por_venta.


* Función `encontrar_columna`para detectar la columna categoría.

In [52]:
def encontrar_columna(df, candidatos):
    for c in df.columns:
        if any(p in c.lower() for p in candidatos):
           return c
    return None

*Detectamos la columna de categoría (acepta variantes)

In [53]:
# 1) Detectar columna de categoría (acepta variantes)
cat_col = encontrar_columna(ventas_clean, ["categoria", "categoría", "categoria_producto", "rubro"])
if cat_col is None:
    raise ValueError("No se encontró columna de categoría (por ej. 'categoria' o 'rubro').")

*Aseguramos columna `ingreso` y si no existe la creamos

In [54]:
# 2) Asegurar columna 'ingreso' (si no existe, crearla)
if "ingreso" not in ventas_clean.columns:
    ventas_cat = ventas_clean.assign(ingreso = ventas_clean["precio"] * ventas_clean["cantidad"])
else:
    ventas_cat = ventas_clean.copy()

* Agregación por categoría con groupby + agg

In [55]:
# 3) Agregación por categoría con groupby + agg
resumen_cat = (
    ventas_cat
    .groupby(
        by=cat_col,
        dropna=False,
        as_index=False
     )
    .agg(
        ingreso_total=('ingreso', 'sum'),
        unidades=('cantidad', 'sum'),
        ventas=('ingreso', 'size'),
        precio_promedio=('precio', 'mean')
    )
    .sort_values(
        by='ingreso_total',
        ascending=False,
        na_position='last',
        ignore_index=True
    )
)

* Ticket promedio por venta = ingreso_total/ventas

In [56]:
# 4) Ticket promedio por venta = ingreso_total / ventas
resumen_cat = resumen_cat.assign(
    ticket_promedio_por_venta = resumen_cat['ingreso_total'] / resumen_cat['ventas']
)

* Imprimimos resultados obtenidos y verificamos.

In [57]:
print("Columna de categoría detectada:", cat_col)
print("Resumen por categoría (ordenado por ingreso_total):")
display(resumen_cat.head(20))


Columna de categoría detectada: categoria
Resumen por categoría (ordenado por ingreso_total):


categoria  ingreso_total  unidades  ventas  precio_promedio  \
0  Electrodomésticos      505299.63      6592    1000         76.52096   
1        Electrónica       482577.8      6413     999         75.25492   
2         Decoración      479216.09      6490    1001         74.09800   

   ticket_promedio_por_venta  
0                  505.29963  
1                 483.060861  
2                 478.737353